algorithms/HPdclassifier/R/hpdegbm.R

```diff
@@ -95,240 +82,292 @@ hpdegbm <- function(
-            train.fraction = NULL,
-            keep.data = FALSE,
-            verbose = FALSE,
-            var.names = NULL,
-            #response.name = "y",
-            group = NULL)
-
-          best.iter0 <- gbm.perf(dGBM_model, 
-              plot.it = FALSE, 
-              oobag.curve = FALSE, 
-              overlay = FALSE, 
-              method="OOB")
-
-          if (best.iter0 <50) best.iter0 <- 50
-
-          best.iter <- as.matrix(best.iter0)
-
-          dGBM_modeli <- list(dGBM_model) 
-  
-          update(dGBM_modeli)
-          update(best.iter)
-        })
+      }
+
+      # Check which classes are in each partition of Y_train. Each partition
+      # must contain all of the classes
+      if (distribution != "gaussian") {
+        npartition_Y <- npartitions(sY_train)
+        # Will store the class names contained in each partition
+        dpartClasses  <- dlist(npartition = npartition_Y) 
+
+        foreach(i, 1:npartition_Y, function(y   = splits(sY_train, i),
+                                            pcs = splits(dpartClasses, i)) {
+          tab <- table(y)
+          # Store names of classes with non-zero count in this partition
+          pcs <- list(names(tab[which(tab != 0)]))
+          update(pcs)
+        }, progress = trace)
+
+        pclasses <- getpartition(dpartClasses)
+
+        if (!all(sapply(pclasses, function(cs) all(cs == pclasses[[1]])))) {
+          if (samplingFlag) 
+            stop("Bad sampling: Some partitions do not contain all classes. Please try again")
+          else 
+            stop("Some partitions do not contain all classes. Try using samplingFlag == TRUE")
+        }
+      }
+
+      # Build local model on each partition of sampled data
+      foreach(i, 1:nExecutor, function(dGBM_modeli       = splits(dl_GBM_model,i), 
+                                       best.iter         = splits(dbest.iter,i), 
+                                       x                 = splits(sX_train, i),
+                                       y                 = splits(sY_train, i),
+                                       n.trees           = n.trees,
+                                       distribution      = distribution,
+                                       interaction.depth = interaction.depth,
+                                       n.minobsinnode    = n.minobsinnode,
+                                       shrinkage         = shrinkage,
+                                       bag.fraction      = bag.fraction,
+                                       buildLocalModel   = .buildLocalGbmModel) {
+        buildLocalModel(dGBM_modeli, best.iter, x, y, n.trees, distribution,
+                        interaction.depth, n.minobsinnode, shrinkage,
+                        bag.fraction)
+      }, progress = trace)
     }
   
     if(trace) {
-	timing_info <- Sys.time() - starttime
-	print(timing_info)
+      timing_info <- Sys.time() - starttime
+      print(timing_info)
```

Comment:
Replaced it in this case. Makes sense. In the function print.hpdegbm, should it also use message? Or would writeLines be more appropriate (same pretty output as message, but written to stdout instead of stderr)?